Specify the HTTP Gateway protocol (#20)
dfinity/interface-spec: spec/index.adoc

lookup_path(["e"],      pruned_tree) = Absent
....

[#http-gateway]
== The HTTP Gateway protocol

This section specifies the _HTTP Gateway protocol_, which allows canisters to handle conventional HTTP requests.

This feature involves the help of a _HTTP Gateway_ that translates between HTTP requests and the IC protocol. Such a gateway could be a stand-alone proxy, it could be implemented in a web browsers (natively, via plugin or via a service worker) or in other ways. This document describes the interface and semantics of this protocol independent of a concrete Gateway, so that all Gateway implementations can be compatible.

> This protocol sits conceptually on top of the interface specified in the remainder of this document, as the involved parties (HTTP Gateway and canisters) are not part of the “system” described here. It is still included in this file due to its role in the Internet Computer ecosystem and due to the importance of keeping multiple Gateway implementations in sync.

Reviewer: This paragraph confused me. I'm not exactly sure what the "system" is. Also, the "it" in "It is still included in this file" didn't make sense to me.

Author: I rewrote it. Better? Or is this not a helpful paragraph?

=== Overview

A HTTP request by an HTTP client is handled by these steps:

1. The Gateway resolves the Host of the request to a canister id.
2. The Gateway Candid-encoded the HTTP request data.
3. The Gateway invokes the canister via a query call to `http_request`.
4. The canister handles the request and returns a HTTP response, encoded in Candid, together with additional metadata.
5. If requested by the canister, the Gateway sends the request again via an update call to `http_request_update`.

> 6. If applicable, the Gateway fetches futher body data via streaming query calls.

Reviewer: Typo: `futher` -> `further`

7. If applicable, the Gateway validates the certificate of the response.
8. The Gateway sends the response to the HTTP client.

[#http-gateway-interface]
=== Candid interface

The following interface description, in https://github.com/dfinity/candid/blob/master/spec/Candid.md[Candid syntax], describes the expecte Canister interface. You can also link:{attachmentsdir}/http-gateway.did[download the file].
----
include::{example}http-gateway.did[]
----

This type signature is a type signature _scheme_, and the canister can choose any type for the token type `T` used for streaming. The Gateway handles that type generically.

Only canisters that use the “Upgrade to update calls” feature need to provide the `http_request_update` method.

> NOTE: Canisters not using these upgrade these features can completeley leave out the `streaming_strategy` and/or `upgade` fields in the `HttpResponse` they return, due to how Candid subtyping works. This might simplifiy their code.

Reviewer: Typo: `Canisters not using these features can completely`

`simplifiy` -> `simplify`

[#http-gateway-name-resolution]
=== Canister resolution

The Gateway needs to know the canister id of the canister to talk to, and obtains that information from the hostname as follows:

1. Check that the hostname, taken from the `Host` field of the HTTP request, is of the form `<name>.raw.ic0.app` or `<name>.ic0.app`, or fail.

2. If the `<name>` is in the following table, use the given canister ids:
+
.Canister hostname resolution
|===
|Hostname |Canister id
|identity | rdmx6-jaaaa-aaaaa-aaadq-cai
|nns | qoctq-giaaa-aaaaa-aaaea-cai
|dscvr | h5aet-waaaa-aaaab-qaamq-cai

> |personhood | g3wsl-eqaaa-aaaan-aaaaa-cai

Reviewer: Where is `personhood` specified?

I don't see it [here](https://github.com/dfinity/agent-js/blob/6e8c64cf07c7722aafbf52351eb0f19fcb954ff0/apps/sw-cert/src/sw/http_request.ts#L15).

Not really convinced this canister resolution belongs in the HTTP Gateway protocol, although it is good to know in general I suppose.

Author: I saw it for the first time yesterday in https://github.com/dfinity/ic/blob/a254116be472f81b85b76429ac8d987f80dd4411/typescript/service-worker/src/sw/http_request.ts#L17. I'll leave it to the Dfinity people to say if this should be there or not. But it demonstrates the need for a authorative list, at least until we have a dynamic system in place.

3. Else, if `<name>` a valid textual encoding of a principal, use that principal as the canister ids.

4. Else fail.

If the hostname was of the form `<name>.ic0.app`, it is a _safe_ hostname; if it was of the form `<name>.raw.ic0.app` it is a _raw_ hostname.

=== Request encoding

The HTTP request is encoded into the `HttpRequest` candid structure.

The `method` field contains the HTTP methd (e.g. `HTTP`), in upper case.

> The `url` field contains the URL from the HTTP request line, i.e. without protocol or hostname, and including query parameters.

Reviewer: Very important! The query parameters inclusion tripped me up.

> The `headers` field contains the headers of the HTTP request.

Reviewer: I wonder if the protocol should specify which request headers are special and will be automatically processed by any Gateway implementation.

I suppose `Host` is an example of one, but how about `Cache-Control` or any others? icx-proxy doesn't do any caching AFAIK (although boundary nodes, which run icx-proxy, apparently do). Should these kind of features be specified in the protocol?

Author: Yes, definitely! Unfortunately I don't know the precise state of affairs here, and hope that @jplevyak can fill in the blanks.

Although maybe some things need not be documented. If the boundary nodes acts on certain headers in a HTTP spec compliant way, e.g. while caching, then that should be fine even if it isn't written here. just as if some user uses their ISP's proxy and that caches.

Reviewer 2: Imho, the gateway should interfere with the HTTP interaction as little as possible. There is one issue though, which is that the service worker is unable to pass on `Set-Cookie` headers, as they are `forbidden`: https://fetch.spec.whatwg.org/#forbidden-response-header-name

Author: This only affects service workers, but not Gateways implemented as proxies or IC-aware browsers, right?

In that case I would not forbid them here, but maybe leave a NOTE that not all Gateway implementations may be able to pass these headers on?

Author: Done. Feel free to suggest edits to the wording if it should be further refined.

The `body` field contains the body of the HTTP request (without any content encodings processed by the Gateway).

=== Upgrade to update calls

> If the canister sets `update = opt true` in the `HttpResponse` reply from `http_request`, then the Gateway ignores all other fields of the reply. The Gateway performs an _update_ call, with the same argument, to `http_request_update`, and uses that response instead.

Reviewer: Probably clearer to say "with the same `HttpRequest`" instead of "with the same argument" IMO

The value of the `update` field returned from `http_request_update` is ignored.

=== Response decoding and streaming

The Gateway assembles the HTTP response from the given `HttpResponse` record.

The HTTP response status code is taken from the `status_code` field.

The HTTP response headers are taken from the `headers` field.

The HTTP response body is initialized with the value of the `body` field.

If the `streaming_strategy` field is set, the HTTP Gateway then uses further query calls to obtain further chunks to append to the body:

1. If the function reference in the `callback` field of the `streaming_strategy` is not a method of the given canister, the Gateway fails the request.

2. Else, it makes a query call to the give method, passing the `token` value given in the `streaming_strategy` as the argument.

> 3. That query method returns a `StreamingCallbackHttpResponse`. The `body` therein is appended to the body of the HTTP response. This is repeated as long as the method returns some token in the `token` field, until that field is `null`.

Reviewer: Slightly off-topic, but I've always been confused exactly how this streaming actually gets implemented. Is it HTTP/1.1 chunked transfer encoding or HTTP/2 streaming?

Should the protocol described in this document talk about whether HTTP/1.1 or HTTP/2 (or both) are supported?

Author: This streaming is _only_ concerned with getting the data from the canister to the Gateway, to overcome the message size limit, so it's independent of any HTTP-level streaming. Note that with certification (as it is specified now), the Gateway can't even stream out the data until it has received it all. In the uncertified case it could stream it out (on the TCP level, or the HTTP level) - I don't care and leave that as as implementation choice. Note that if the Gateway is the browser itself it also doesn't make sense to stream.

But it's a good point and I'll clarify!

> === Response certification

Reviewer: I don't feel qualified to review the certification part at the moment, so I'll defer this for someone else.

If the hostname was of the form `<name>.ic0.app`, it is a _safe_ hostname; if it was of the form `<name>.raw.ic0.app` it is a _raw_ hostname.

=== Request encoding

The HTTP request is encoded into the `HttpRequest` candid structure.

The `method` field contains the HTTP methd (e.g. `HTTP`), in upper case.

The `url` field contains the URL from the HTTP request line, i.e. without protocol or hostname, and including query parameters.

The `headers` field contains the headers of the HTTP request.

The `body` field contains the body of the HTTP request (without any content encodings processed by the Gateway).

=== Upgrade to update calls

If the canister sets `update = opt true` in the `HttpResponse` reply from `http_request`, then the Gateway ignores all other fields of the reply. The Gateway performs an _update_ call to `http_request_update`, passing the same `HttpRequest` record as the argument, and uses that response instead.

The value of the `update` field returned from `http_request_update` is ignored.

=== Response decoding

The Gateway assembles the HTTP response from the given `HttpResponse` record:

* The HTTP response status code is taken from the `status_code` field.

> * The HTTP response headers are taken from the `headers` field.

Author: Is the boundary node mangling or filtering the headers in any way that should be noted here?

Reviewer 3: The filtering we have is in [dfinity/ic/ic-os/boundary-guestos/rootfs/etc/nginx/conf.d/001-ic-nginx.conf](https://github.com/dfinity/ic/blob/master/ic-os/boundary-guestos/rootfs/etc/nginx/conf.d/001-ic-nginx.conf#L154).

The short of it is we add these response headers (and thus their implications) to every HTTP call response:

```access-control-allow-credentials: true
access-control-allow-origin: *
access-control-allow-methods: GET, POST, HEAD, OPTIONS
access-control-allow-headers: DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Cookie
access-control-expose-headers: Content-Length,Content-Range
x-cache-status: MISS
```

Author: I guess the question is: Why set these? Shouldn’t the canister have control over them?

Reviewer 3: The short answer is "it's complicated". The longer answer is "because headers are not yet certified". As such we currently limit headers a bit to improve security. Once we get certified headers, we should not have any more need for this limitation.

Author: Ok, good enough for this PR I guess.

=== Overview

A HTTP request by an HTTP client is handled by these steps:

1. The Gateway resolves the Host of the request to a canister id.
2. The Gateway Candid-encoded the HTTP request data.
3. The Gateway invokes the canister via a query call to `http_request`.
4. The canister handles the request and returns a HTTP response, encoded in Candid, together with additional metadata.
5. If requested by the canister, the Gateway sends the request again via an update call to `http_request_update`.
6. If applicable, the Gateway fetches further body data via streaming query calls.
7. If applicable, the Gateway validates the certificate of the response.
8. The Gateway sends the response to the HTTP client.

[#http-gateway-interface]
=== Candid interface

> The following interface description, in https://github.com/dfinity/candid/blob/master/spec/Candid.md[Candid syntax], describes the expecte Canister interface. You can also link:{attachmentsdir}/http-gateway.did[download the file].

Reviewer 4: ```suggestion
The following interface description, in https://github.com/dfinity/candid/blob/master/spec/Candid.md[Candid syntax], describes the expected Canister interface. You can also link:{attachmentsdir}/http-gateway.did[download the file].
```

----
include::{example}http-gateway.did[]
----

Only canisters that use the “Upgrade to update calls” feature need to provide the `http_request_update` method.

> NOTE: Canisters not using these features can completely leave out the `streaming_strategy` and/or `upgade` fields in the `HttpResponse` they return, due to how Candid subtyping works. This might simplify their code.

Reviewer 4: ```suggestion
NOTE: Canisters not using these features can completely leave out the `streaming_strategy` and/or `upgrade` fields in the `HttpResponse` they return, due to how Candid subtyping works. This might simplify their code.
```

[#http-gateway-name-resolution]
=== Canister resolution

The Gateway needs to know the canister id of the canister to talk to, and obtains that information from the hostname as follows:

1. Check that the hostname, taken from the `Host` field of the HTTP request, is of the form `<name>.raw.ic0.app` or `<name>.ic0.app`, or fail.

2. If the `<name>` is in the following table, use the given canister ids:
+
.Canister hostname resolution
|============================================
| Hostname     | Canister id
| `identity`   | `rdmx6-jaaaa-aaaaa-aaadq-cai`
| `nns`        | `qoctq-giaaa-aaaaa-aaaea-cai`
| `dscvr`      | `h5aet-waaaa-aaaab-qaamq-cai`
| `personhood` | `g3wsl-eqaaa-aaaan-aaaaa-cai`
|============================================

> 3. Else, if `<name>` a valid textual encoding of a principal, use that principal as the canister ids.

Reviewer 4: ```suggestion
3. Else, if `<name>` is a valid textual encoding of a principal, use that principal as the canister id.
```

4. Else fail.

If the hostname was of the form `<name>.ic0.app`, it is a _safe_ hostname; if it was of the form `<name>.raw.ic0.app` it is a _raw_ hostname.

=== Request encoding

> The HTTP request is encoded into the `HttpRequest` candid structure.

Reviewer 4: ```suggestion
The HTTP request is encoded into the `HttpRequest` Candid structure.
```

> * The `method` field contains the HTTP methd (e.g. `HTTP`), in upper case.

Reviewer 4: ```suggestion
* The `method` field contains the HTTP method (e.g. `HTTP`), in upper case.
```

* The `url` field contains the URL from the HTTP request line, i.e. without protocol or hostname, and including query parameters.

* The `headers` field contains the headers of the HTTP request.

* The `body` field contains the body of the HTTP request (without any content encodings processed by the Gateway).

=== Upgrade to update calls

If the canister sets `update = opt true` in the `HttpResponse` reply from `http_request`, then the Gateway ignores all other fields of the reply. The Gateway performs an _update_ call to `http_request_update`, passing the same `HttpRequest` record as the argument, and uses that response instead.

The value of the `update` field returned from `http_request_update` is ignored.

=== Response decoding

The Gateway assembles the HTTP response from the given `HttpResponse` record:

* The HTTP response status code is taken from the `status_code` field.

* The HTTP response headers are taken from the `headers` field.
+

> NOTE: Not all Gateway implementations may be able to pass on all forms of headers. In particular, Service Workers are unable to pass on the `Set-cookie` header.

Reviewer 4: ```suggestion
NOTE: Not all Gateway implementations may be able to pass on all forms of headers. In particular, Service Workers are unable to pass on the `Set-Cookie` header.
```

+
[NOTE]
====

> HTTP Gateway may add addditional headers. In particular, the following headers may be set:

Reviewer 4: ```suggestion
HTTP Gateways may add additional headers. In particular, the following headers may be set:
```

....
access-control-allow-origin: *
access-control-allow-methods: GET, POST, HEAD, OPTIONS
access-control-allow-headers: DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Cookie
access-control-expose-headers: Content-Length,Content-Range
x-cache-status: MISS
....
====

* The HTTP response body is initialized with the value of the `body` field, and further assembled as per the <<http-gateway-streaming,streaming protocol>>.

[#http-gateway-streaming]
=== Response body streaming

> The HTTP Gateway protocol has provisions transfer further chunks of the body data from the canister to the HTTP Gateway, to overcome the message limit of the Internet Computer. This streaming protocol is idependent of any possible streaming of data between the HTTP Gateway and the HTTP client. The gateway may assemble the response in whole before passing it on, or pass the chunks on directly, on the TCP or HTTP level, as it sees fit. When the Gateway is <<http-gateway-certification,certifying the response>>, it must not pass on uncertified chunks.

Reviewer 4: ```suggestion
The HTTP Gateway protocol has provisions to transfer further chunks of the body data from the canister to the HTTP Gateway, to overcome the message limit of the Internet Computer. This streaming protocol is independent of any possible streaming of data between the HTTP Gateway and the HTTP client. The gateway may assemble the response in whole before passing it on, or pass the chunks on directly, on the TCP or HTTP level, as it sees fit. When the Gateway is <<http-gateway-certification,certifying the response>>, it must not pass on uncertified chunks.
```

If the `streaming_strategy` field of the `HttpResponse` is set, the HTTP Gateway then uses further query calls to obtain further chunks to append to the body:

1. If the function reference in the `callback` field of the `streaming_strategy` is not a method of the given canister, the Gateway fails the request.

> 2. Else, it makes a query call to the give method, passing the `token` value given in the `streaming_strategy` as the argument.

Reviewer 4: ```suggestion
2. Else, it makes a query call to the given method, passing the `token` value given in the `streaming_strategy` as the argument.
```

3. That query method returns a `StreamingCallbackHttpResponse`. The `body` therein is appended to the body of the HTTP response. This is repeated as long as the method returns some token in the `token` field, until that field is `null`.

> WARNING: The type of the `token` value is chosen by the canister; the HTTP Gateway obtains the Candid type from the encoded message from the canister, and use it when passing the token back to the canister. This generic use of Candid is no covered by the Candid specification, and may not be possible in some cases (e.g. when using “future types”). Canister authors may have to use “simple” types.

Reviewer 4: ```suggestion
WARNING: The type of the `token` value is chosen by the canister; the HTTP Gateway obtains the Candid type of the encoded message from the canister, and uses it when passing the token back to the canister. This generic use of Candid is not covered by the Candid specification, and may not be possible in some cases (e.g. when using “future types”). Canister authors may have to use “simple” types.
```

[#http-gateway-certification]
=== Response certification

If the hostname was safe, the HTTP Gateway performs _certificate validation_:

1. It searches for a response header called `Ic-Certificate` (case-insensitive).

> 2. The value of the header must be a structured header according to RFC8941 with fields `certificate` and `tree`, both being byte sequences.

Reviewer 4: ```suggestion
2. The value of the header must be a structured header according to RFC 8941 with fields `certificate` and `tree`, both being byte sequences.
```

> 3. The `certificate` must be a valid certificate as per <<certification>>, signed by the root key. If the certificate contains a subnet delegation, the delegation must be valid for the given canister. The timestamp in `/time` must be recent. The subnet state tree in the certificate must reveal the canister’s  <<state-tree-certified-data,certified data>>.

Reviewer 4: ```suggestion
3. The `certificate` must be a valid certificate as per <<certification>>, signed by the root key. If the certificate contains a subnet delegation, the delegation must be valid for the given canister. The timestamp in `/time` must be recent. The subnet state tree in the certificate must reveal the canister’s <<state-tree-certified-data,certified data>>.
```

> 3. The `certificate` must be a valid certificate as per <<certification>>, signed by the root key. If the certificate contains a subnet delegation, the delegation must be valid for the given canister. The timestamp in `/time` must be recent. The subnet state tree in the certificate must reveal the canister’s <<state-tree-certified-data,certified data>>.

Reviewer 2: Currently neither icx-proxy nor the service worker check the certificate `/time` (which is bad). Can you give some more specific measure of "recent"? 5 minutes?

Author: That’s probably for the security team to decide. 5 min should be ample (until we get into edge node caching, then things become tricky, because a cache is hardly different from an attacker serving old data).

> 4. The `tree` must be a hash trees as per <<certification-encoding>>.

Reviewer 4: ```suggestion
4. The `tree` must be a hash tree as per <<certification-encoding>>.
```

5. The root hash of that `tree` must match the canister’s certified data.

6. The path `["http_assets",<url>]`, where `url` is the utf8-encoded `url` from the `HttpRequest` must exist and be a leaf.  Else, if it does not exist, `["http_assets","/index.html"]` must exist and be a leaf.

> 7. That leaf must contain the SHA256 of the _decoded_ body.

Reviewer 4: ```suggestion
7. That leaf must contain the SHA-256 hash of the _decoded_ body.
```
